File: src/CourseRegistrationService.java

import java.util.ArrayList;

public class CourseRegistrationService {
    private ArrayList<CourseRequest> _courseRequests = new ArrayList<CourseRequest>();

    public void createCourseRequest(Student student, Course course) {
        _courseRequests.add(new CourseRequest(student, course, student.get_advisorID()));

    }

    public ArrayList<CourseRequest> checkAccesiableRequests(Advisor advisor) {
        ArrayList<CourseRequest> accessibleRequests = new ArrayList<>();

        for (CourseRequest courseRequest : _courseRequests) {
            if (courseRequest.get_advisorID().equals(advisor.get_staffId())) {
                accessibleRequests.add(courseRequest);
            }
        }

        return accessibleRequests;
    }

    public void removeCourseRequest(CourseRequest courseRequest) {
        _courseRequests.remove(courseRequest);
    }
}

class CourseRequest {
    private final Student _student;
    private final Course _course;
    private final StaffId _advisorID;

    public CourseRequest(Student student, Course course, StaffId advisorID) {
        _student = student;
        _course = course;
        _advisorID = advisorID;
    }

    public Student get_student() {
        return _student;
    }

    public Course get_course() {
        return _course;
    }

    public StaffId get_advisorID() {
        return _advisorID;
    }
}

File: pythonCode/CourseRegistrationService.py
import logging

from pythonCode.Logger import setup_logger
logger = setup_logger("CourseRegistrationService")
class CourseRegistrationService:
    def __init__(self):
        logging.getLogger().handlers.clear()
        logger.info(f"{self.__class__.__name__} classes created.")
        self._courseRequests = list()

    def createCourseRequest(self, student, course):
        logger.info("Creating a course request.")
        self._courseRequests.append(CourseRequest(student, course, student.get_advisorID()))

    def checkAccesiableRequests(self, advisor):
        logger.info("Checking accessible requests.")
        accessibleRequests = list()
        for courseRequest in self._courseRequests:
            if advisor.get_staffId().get_staff_id() == courseRequest.get_advisorID().get_staff_id():
                accessibleRequests.append(courseRequest)
        return accessibleRequests

    def removeCourseRequest(self, courseRequest):
        logger.info("Removing a course request.")
        self._courseRequests.remove(courseRequest)

    def checkClassroomAvailabilityForDayAndTime(self, day, time, classroom, allCourses):
        logger.info(f"Checking classroom availability {day} {time} {classroom.get_classroom_name()}")
        for course in allCourses:
            for courseSection in course.getCourseSections():
                if courseSection._day == day and courseSection._sectionTime == time and courseSection._classroom == classroom:
                    return False
        return True


class CourseRequest:
    def __init__(self, student, course, advisorID):
        logging.getLogger().handlers.clear()
        logger.info(f"{self.__class__.__name__} classes created.")
        self._student = student
        self._course = course
        self._advisorID = advisorID

    def get_student(self):
        logger.info(f"The get_student method in the {self.__class__.__name__} class is called.")
        return self._student

    def get_course(self):
        logger.info(f"The get_course method in the {self.__class__.__name__} class is called.")
        return self._course

    def get_advisorID(self):
        logger.info(f"The get_advisorID method in the {self.__class__.__name__} class is called.")
        return self._advisorID
